sql/src/main/java/io/crate/expression/scalar/cast/CastFunction.java

```diff
@@ -28,26 +28,119 @@
 import io.crate.expression.symbol.Function;
 import io.crate.expression.symbol.Literal;
 import io.crate.expression.symbol.Symbol;
-import io.crate.metadata.BaseFunctionResolver;
 import io.crate.metadata.FunctionIdent;
-import io.crate.metadata.FunctionImplementation;
 import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.FunctionName;
 import io.crate.metadata.Scalar;
 import io.crate.metadata.TransactionContext;
-import io.crate.metadata.functions.params.FuncParams;
-import io.crate.metadata.functions.params.Param;
+import io.crate.metadata.functions.Signature;
 import io.crate.types.DataType;
-import io.crate.types.UndefinedType;
 
-import java.util.List;
 import java.util.Map;
 import java.util.function.BiFunction;
 
 import static io.crate.expression.scalar.cast.CastFunctionResolver.CAST_SIGNATURES;
 import static io.crate.expression.scalar.cast.CastFunctionResolver.TRY_CAST_PREFIX;
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
 
 public class CastFunction extends Scalar<Object, Object> {
 
+    public static void register(ScalarFunctionModule module) {
+        // We still maintain the cast function to type mapping to stay
+        // bwc by keeping the old `to_<type>` and `try_<type>` function signatures.
+        //
+        // We can drop the per type cast function already in 4.2. This change
+        // would require handling the metadata for the places where the old
+        // signature of the cast function were used, e.g. generated columns.
+        for (Map.Entry<String, DataType> function : CAST_SIGNATURES.entrySet()) {
+            module.register(
+                Signature.builder()
+                    .name(new FunctionName(null, function.getKey()))
+                    .kind(FunctionInfo.Type.SCALAR)
+                    .typeVariableConstraints(typeVariable("E"), typeVariable("V"))
+                    .argumentTypes(parseTypeSignature("E"), parseTypeSignature("V"))
+                    .returnType(function.getValue().getTypeSignature())
+                    .forbidCoercion()
+                    .build(),
+                args -> {
+                    DataType<?> sourceType = args.get(0);
+                    DataType<?> targetType = args.get(1);
+                    if (!sourceType.isConvertableTo(targetType)) {
+                        throw new ConversionException(sourceType, targetType);
+                    }
```

Comment:
this check can be removed, the signature binder will already take care of this (if not its a bug ;-).

Reply:
hm looked at it a bit, I am not sure whether it is in the scope of the signature binding. It looks like it is part of the cast logic, checking whether the first argument of the function can be converted to the second. Maybe it can be even moved into the CastFunctionReolver and checked based in tryCast flag.

Reply:
Ah right, these are 2 different type variables.